### ISSUE TITLE
Unable to create sandbox on NFS volume with apptainer, while it was possible with singularity.

### ISSUE THREAD

**Issue by Aneoshun** [apptainer/apptainer#1188]
### Version of Apptainer

What version of Apptainer (or Singularity) are you using?

`apptainer version 1.1.3`

### Expected behavior

I recently migrated from Singularity 3.4.1 to Apptainer 1.1.3. My daily usage was to build and use sandbox containers using --fakeroot (setuid) on a NFS location. Which worked very well so far with singularity, but does not work with Apptainer. I have seen the documentation, so I understand that it is the expected behaviour. I am just surprised that I am just unable to do something that worked before.

Do you know a workaround, or what would be an alternative network-hosted filesystem that can be used without issues with apptainer and sandboxes?
To be more concrete, I host the home directories of a handful of researchers on a Synology NAS that are mounted automatically via NFS. I will be happy to explore alternative solutions, but I am not familiar with this. 

### Actual behavior

The error I get is the following: 
````
INFO    [U=0,P=1]          Assemble()                    Creating sandbox directory...
DEBUG   [U=0,P=1]          Assemble()                    Copying sandbox from /tmp/build-temp-2542497623/rootfs to /home_nfs/XXX/XXX/XXX/XXX.sif
DEBUG   [U=0,P=1]          cleanUp()                     Cleaning up "/tmp/build-temp-2542497623/rootfs" and "/tmp/bundle-temp-2576191350"
ERRO[0491] Can't add file /tmp/build-temp-2542497623/rootfs/etc/ld.so.cache to tar: io: read/write on closed pipe
ERRO[0491] Can't close tar writer: io: read/write on closed pipe
FATAL   [U=0,P=1]          runBuildLocal()               While performing build: copy Failed: lchown /home_nfs/antoine/git/mapelites-example/singularity/mapelites-example.sif/etc/gshadow: operation not permitted
DEBUG   [U=0,P=1651062]    Master()                      Child exited with exit status 255 
````

### Steps to reproduce this behavior
I am running Apptainer 1.1.3, installed locally, with SINGULARITY_CACHEDIR and TMP on a local drive. I do not know if this is enough to reproduce

### What OS/distro are you running

`Ubuntu 22.04.1 LTS`

### How did you install Apptainer
From the deb file downloaded from github.

**Comment by DrDaveD** [comment]
You have /etc/subuid and /etc/subgid mappings in place, correct?  Do you get different behavior if you either comment out your entry in /etc/subuid or add the (undocument) option `--ignore-subuid` so it falls back to using the fakeroot command?

It would probably be helpful if you included the container definition file you are trying to build.

**Comment by DrDaveD** [comment]
Also it appears that it is not putting the temporary files on a local drive, which might be a problem.  Setting SINGULARITY_TMPDIR or APPTAINER_TMPDIR instead of TMP help with that.

**Comment by Aneoshun** [comment]
Hi @DrDaveD, 
Thanks for your inputs. 

For all the examples below, I set SINGULARITY_TMPDIR to be on the local drive (non-nfs). 
I found a minimal singularity def file that raises the error: 

```
BootStrap: docker
From: python

%help
   install python

%post
apt update
DEBIAN_FRONTEND=noninteractive apt install -y --no-install-recommends emacs

%environment


%runscript

%labels
   Author me
```

If I do: `apptainer build --fakeroot --sandbox mysandbox singularity.def`
I get:
```
INFO:    Creating sandbox directory...
ERRO[0039] Can't add file /.singularity/antoine/build-temp-4145099171/rootfs/usr/lib/emacs/27.1/x86_64-linux-gnu/emacs.pdmp to tar: io: read/write on closed pipe
ERRO[0039] Can't close tar writer: io: read/write on closed pipe
FATAL:   While performing build: copy Failed: lchown /home_nfs/antoine/test_apptainer/mysandbox/usr/lib/dbus-1.0/dbus-daemon-launch-helper: operation not permitted
```


If I do: `apptainer build --fakeroot --sandbox --ignore-subuid mysandbox2 singularity.def`
I get a strange error about a missing library and it takes ages:
```
INFO:    Running post scriptlet
WARNING: DEPRECATED USAGE: Environment variable SINGULARITY_TMPDIR will not be supported in the future, use APPTAINER_TMPDIR instead
/.singularity.d/libs/faked: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.33' not found (required by /.singularity.d/libs/faked)
/.singularity.d/libs/faked: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.34' not found (required by /.singularity.d/libs/faked)
fakeroot: error while starting the `faked' daemon.
/.singularity.d/libs/fakeroot: 1: kill: Usage: kill [-s sigspec | -signum | -sigspec] [pid | job]... or
kill -l [exitstatus]
FATAL:   While performing build: while running engine: exit status 1
```

FInally, if I do `singularity_3.4.1 build --fakeroot --sandbox  mysandbox3 singularity.def` it just works. 

Also, if I do something else than `apt install emacs` that does not require sudo, like `pip install numpy`, it actually works with apptainer.

Any thoughts?

**Comment by DrDaveD** [comment]
When I try to build with your definition file on a CentOS7 host and NFS destination, it starts with this warning:
```
WARNING: The underlying filesystem on which resides "/nashome/d/dwd/scratch/issue1188" won't allow to set ownership, as a consequence the sandbox could not preserve image's files/directories ownerships
```
Do you get that?  In my case it ends with a different error but it is still `operation not permitted`:
```
FATAL:   While performing build: copy Failed: chown /nashome/d/dwd/scratch/issue1188: operation not permitted
```
So I think it's just not going to work to do a build with any part on NFS.  It works for me to store a SIF file there, when not using `--sandbox` and when the APPTAINER_TMPDIR is a local filesystem.  When you build with `--sandbox` it tries to build the container right there at that destination, but if you're building a SIF then it builds the container in the temporary directory first, then squashes it into the SIF file.  The latter works but not the former.

The error you get with `--ignore-subuid` is a different issue, a known problem.  Ubuntu 22.04 uses a version of glibc that is incompatible with most other currently supported host operating systems, which causes the fakeroot command to fail.   So the /etc/subuid fakeroot should in general be better for you.  However, it could still be interesting to see if you can avoid the lchown error with that type of fakeroot if you switched to building a container based on Ubuntu 22.04, because then that kind of fakeroot should work for a build ... oh, but I just tried it on my CentOS7 host and no, I still got an operation not permitted error:
```
FATAL:   While performing build: packer failed to pack: while unpacking tmpfs: error unpacking rootfs: chown rootfs: lchown /nashome/d/dwd/scratch/build-temp-2724263889/rootfs: operation not permitted
```
So I don't think there's a workaround to this limitation on NFS, other than keeping APPTAINER_TMPDIR off of NFS and avoiding building sandboxes on NFS.

**Comment by Aneoshun** [comment]
Yes, I get the same warning, and it is explicitly said in the documentation that Sandboxes do work on NFS. So, I am not too surprised that there is no workaround. 

However, I am still curious to know why it was possible with Singularity (e.g., 3.4.1), but not anymore. 

Also, would you have any recommendations for an alternative FS compatible with Apptainer for a shared volume online?

Thanks!

**Comment by DrDaveD** [comment]
I assume you meant "do not work on NFS", not "do work on NFS".

I tried singularity 3.4.2 and noticed that it built the sandbox in $SINGULARITY_TMPDIR, then copied it out to the final destination.  That method would have avoided the issue but would have been a lot slower.  There must have been a later optimization that skipped the temporary directory.  I verified that version 3.5.3 does not use a temporary directory (I don't have easy access to earlier 3.5.x binaries).  There is some discussion in the 3.5.x release announcements about the issue, although I didn't narrow it down to where and why exactly the behavior changed.

I am closing the issue.  Reopen it if you think there's something still to be done.
